libminifi/include/FlowController.h

```diff
@@ -252,8 +255,7 @@ class FlowController : public core::controller::ForwardingControllerServiceProvi
   // Thread pool for schedulers
   utils::ThreadPool<utils::TaskRescheduleInfo> thread_pool_;
   std::map<utils::Identifier, std::unique_ptr<state::ProcessorController>> processor_to_controller_;
+  std::unique_ptr<state::MetricsPublisher> metrics_publisher_;
```

Comment:
The flow controller shouldn't own this service, or even depend on it. Each entity that's capable of exporting metrics should expose that on its public interface. The flow controller may provide some aggregation into a single structure if necessary, but it shouldn't do much else.
Let's have the MetricsPublisher depend on a metric source callback, that's bound to this aggregation member function of the flow controller, without each depending on the other.